### PR TITLE
Add functional javascript tests

### PR DESCRIPTION
**Pull request by bboro**
#307

### CODE REVIEW

**Comment by bboro** [comment]
TB: 1.7 / 1h (Total 10.7 / 10h)

File: web/modules/custom/server_general/tests/src/ExistingSite/ServerGeneralSelenium2TestBase.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+/**
+ * Base test class for writing ExistingSite tests with JS capability.
+ */
+class ServerGeneralSelenium2TestBase extends ExistingSiteSelenium2DriverTestBase {
+
+  /**
+   * Take a screenshot and save it in sites/simpletest/screenshots.
+   *
+   * Note: Do not leave a call to this method in commits. You should only
+   * take screenshots locally for debugging purposes.
```
Comment (amitaibu) [inline web/modules/custom/server_general/tests/src/ExistingSite/ServerGeneralSelenium2TestBase.php:18]: :+1: 
Let's add an entry in the README, and mention this debug method

Comment (mariano-dagostino) [reply web/modules/custom/server_general/tests/src/ExistingSite/ServerGeneralSelenium2TestBase.php:18]: @bboro What about checking if env CI or, maybe expect a local ENV variable to take an screenshot?

File: README.md [hunk 2]
```diff
@@ -29,6 +29,7 @@ deployments. See more under ["Deploy to Pantheon"](#deploy-to-pantheon) section
 ## Installation
 
     ddev composer install
+    ddev get ddev/ddev-selenium-standalone-chrome
```
Comment (mariano-dagostino) [inline README.md:32]: Do we need this step?

**Review by mariano-dagostino — COMMENTED** [review]
:clap: :clap: :clap: :clap:

File: README.md (hunk 2 above)
Comment (bboro) [reply README.md:32]: If the tests pass now, yes we need this step :)

File: web/modules/custom/server_general/tests/src/ExistingSite/ServerGeneralSelenium2TestBase.php (hunk 1 above)
Comment (bboro) [reply web/modules/custom/server_general/tests/src/ExistingSite/ServerGeneralSelenium2TestBase.php:18]: Thanks @mariano-dagostino updated to exclude CI envs

**Comment by bboro** [comment]
Server General Homepage (Drupal\Tests\server_general\ExistingSite\ServerGeneralHomepage)

 ✘ Homepage cache

   │

   │ Behat\Mink\Exception\UnsupportedDriverActionException: Response headers are not available from Behat\Mink\Driver\Selenium2Driver


Ok I'm gonna have to split these tests into 2

**Comment by amitaibu** [comment]
Bump to total 4h

**Comment by bboro** [comment]
@amitaibu this is completed btw.

File: web/modules/custom/server_general/tests/src/ExistingSite/ServerGeneralHomepageTest.php [hunk 3]
```diff
@@ -2,21 +2,36 @@
...
+    $this->assertNotNull($carousel);
+
+    // Assert the current slide has the expected title.
+    $web_assert->elementTextContains('css', '.paragraph--type--related-content .carousel-slide.slick-current', 'Current Digital Marketing Is Sports-Watching, Rather Than Marketing');
+    // Click on the 2nd dot navigation.
+    $dots_2 = $carousel->find('css', '.slick-dots li:nth-child(2)');
+    $this->assertNotNull($dots_2);
+    $dots_2->click();
+    // Wait 1 sec for the JS and animation.
+    $this->getSession()->wait(1000);
```
Comment (amitaibu) [inline web/modules/custom/server_general/tests/src/ExistingSite/ServerGeneralHomepageTest.php:32]: I think there's a wait with function, so we could wait for the slide to change

File: README.md [hunk 4]
```diff
@@ -182,6 +183,14 @@ See the [example](https://github.com/Gizra/drupal8-starter/blob/master/web/modul
     # Run a single method from a test file.
     ddev phpunit --filter testHomepageCache web/modules/custom/server_general/tests/src/ExistingSite/ServerGeneralHomepageTest.php
 
+We also have capability to write tests which run on a headless chrome browser with
+Javascript capabilities. See `Drupal\Tests\server_general\ExistingSite\ServerGeneralSelenium2TestBase`
```
Comment (amitaibu) [inline README.md:187]: Let's have a link to the GH file, So `See [ServerGeneralSelenium2TestBase](...) for ...`

File: README.md (hunk 2 above)
Comment (AronNovak) [reply README.md:32]: https://ddev.readthedocs.io/en/latest/users/usage/commands/#get - `ddev get` downloads an add-on. But this project already contains this add-on, it happened while creating this PR.

File: ci-scripts/install_ddev.sh [hunk 5]
```diff
@@ -17,3 +17,5 @@ cp "ci-scripts/global_config.yaml" ~/.ddev/
 docker network create ddev_default || ddev logs
 
 ddev composer install || ddev logs
+
+ddev get ddev/ddev-selenium-standalone-chrome || ddev logs
```
Comment (AronNovak) [inline ci-scripts/install_ddev.sh:21]: I doubt if we need this.

File: web/modules/custom/server_general/tests/src/ExistingSite/ServerGeneralHomepageTest.php (hunk 3 above)
Comment (bboro) [reply web/modules/custom/server_general/tests/src/ExistingSite/ServerGeneralHomepageTest.php:32]: Yes but I couldn't figure out what's needed in the $conditions array. It seems like a pain to have to write the condition in JS. Easier to just wait 1sec (300ms also works here as that's the animation duration, but thought 1000ms to be safer)

File: ci-scripts/install_ddev.sh (hunk 5 above)
Comment (bboro) [reply ci-scripts/install_ddev.sh:21]: Tests wouldn't run without this, not sure why.

Comment (bboro) [reply ci-scripts/install_ddev.sh:21]: Correction: Tests actually ran, but they fail saying can't connect to port 4444 on the JS tests

File: web/modules/custom/server_general/tests/src/ExistingSite/ServerGeneralHomepageTest.php (hunk 3 above)
Comment (bboro) [reply web/modules/custom/server_general/tests/src/ExistingSite/ServerGeneralHomepageTest.php]: I've reduced the wait time from 1000 to 500

File: ci-scripts/install_ddev.sh (hunk 5 above)
Comment (bboro) [reply ci-scripts/install_ddev.sh]: Yep so we do need this line.

See: https://app.travis-ci.com/github/Gizra/drupal-starter/jobs/605250363#L1752 for below commit [Try to run test without ddev get](https://github.com/Gizra/drupal-starter/pull/521/commits/aca4c59c58448bdf94b0c7279fa5c2ad5cc4b5ab)

**Comment by bboro** [comment]
TB: 3.5 / 4h

File: README.md (hunk 2 above)
Comment (bboro) [reply README.md:32]: @AronNovak I think this is most likely due to the fact that we're not committing all the artefacts that get installed when we run this.

File: ci-scripts/install_ddev.sh (hunk 5 above)
Comment (AronNovak) [reply ci-scripts/install_ddev.sh:21]: I pushed the missing file, could we please try without this install step? Now it should work.
My concern is the following. We do not freeze version number, we have frozen half of the stuff in the repo, half of the stuff would come from the repo, the latest version.

Comment (bboro) [reply ci-scripts/install_ddev.sh:21]: Sure thanks. Tests pass still

File: README.md (hunk 2 above)
Comment (AronNovak) [inline README.md:32]:
```suggestion
```

**Pull request merged**